website/blog/2025/02-12-scrape-bluesky-using-python/index.md

```diff
@@ -0,0 +1,721 @@
+
+        await context.add_requests([str(next_url)])
+```
+
+When receiving user data, we'll store it in the corresponding Dataset `self._users`
+
+```python
+async def _user_handler(self, context: HttpCrawlingContext) -> None:
+    context.log.info(f'Processing user {context.request.url} ...')
+
+    data = json.loads(context.http_response.read())
+
+    user_item = {
+        'did': data['did'],
+        'created': data['createdAt'],
+        'avatar': data.get('avatar'),
+        'description': data.get('description'),
+        'display_name': data.get('displayName'),
+        'handle': data['handle'],
+        'indexed': data.get('indexedAt'),
+        'posts_count': data['postsCount'],
+        'followers_count': data['followersCount'],
+        'follows_count': data['followsCount'],
+    }
+
+    await self._users.push_data(user_item)
+```
+
+### 5. Saving data to files
+
+For saving results, we'll use the `write_to_json` method in Dataset.
+
+```python
+async def save_data(self) -> None:
+    """Save the data."""
+    if not self._users or not self._posts:
+        raise ValueError('Datasets not initialized.')
+
+    with open('users.json', 'w') as f:
+        await self._users.write_to_json(f, indent=4)
+
+    with open('posts.json', 'w') as f:
+        await self._posts.write_to_json(f, indent=4)
+```
+
+### 6. Running the crawler
+
+We have everything needed to complete the crawler. We just need a method to execute the crawling - let's call it `crawl`
+
+```python
+async def crawl(self, queries: list[str]) -> None:
+    """Crawl the given URL."""
+    if not self._crawler:
+        raise ValueError('Crawler not initialized.')
+
+    await self._crawler.run([f'{self._domain}/xrpc/app.bsky.feed.searchPosts?q={query}' for query in queries])
+```
+
+Let's finalize the code:
+
+```python
+async def run() -> None:
+    """Main execution function that orchestrates the crawling process.
+
+    Creates a crawler instance, manages the session, and handles the complete
+    crawling lifecycle including proper cleanup on completion or error.
+    """
+    crawler = BlueskyCrawler()
+    crawler.create_session()
+    try:
+        await crawler.init_crawler()
+        await crawler.crawl(['python', 'apify', 'crawlee'])
+        await crawler.save_data()
+    except Exception:
+        traceback.print_exc()
+    finally:
+        crawler.delete_session()
+
+
+def main() -> None:
+    """Entry point for the crawler application."""
+    asyncio.run(run())
+```
+
+If you check your `pyproject.toml`, you'll see that UV created an entrypoint for running `bluesky-crawlee = "bluesky_crawlee:main"`, so we can run our crawler simply by executing:
+
+```bash
+uv run bluesky-crawlee
+```
+
+Let's look at sample results:
+
+Posts
+
+![Posts Example](./img/posts.webp)
+
+Users
+
+![Users Example](./img/users.webp)
+
+## Create Apify actor for Bluesky crawler
```

Comment:
link to what is an Actor, and a little explanation why are we making Actor, because its the easiest way to deploy a software on cloud, etc, etc.

also loved it :)